akka-http-core/src/main/scala/akka/http/impl/engine/http2/client/PersistentConnection.scala

```diff
@@ -53,22 +61,77 @@ private[http2] object PersistentConnection {
             pull(requestIn)
-
       }
 
-      def connect(): Unit = {
+      def connect(connectsLeft: Option[Int]): Unit = {
         val requestOut = new SubSourceOutlet[HttpRequest]("PersistentConnection.requestOut")
         val responseIn = new SubSinkInlet[HttpResponse]("PersistentConnection.responseIn")
+        val connection = Promise[OutgoingConnection]()
 
-        Source.fromGraph(requestOut.source)
-          .via(connectionFlow)
-          .runWith(responseIn.sink)(subFusingMaterializer)
+        become(new Connecting(connection.future, requestOut, responseIn, connectsLeft.map(_ - 1)))
 
-        become(new Connected(requestOut, responseIn))
+        connection.completeWith(Source.fromGraph(requestOut.source)
+          .viaMat(connectionFlow)(Keep.right)
+          .toMat(responseIn.sink)(Keep.left)
+          .run()(subFusingMaterializer))
+      }
+
+      class Connecting(
+        connected:    Future[OutgoingConnection],
+        requestOut:   SubSourceOutlet[HttpRequest],
+        responseIn:   SubSinkInlet[HttpResponse],
+        connectsLeft: Option[Int]
+      ) extends State {
+        connected.onComplete({
+          case Success(_) =>
+            onConnected.invoke(())
+          case Failure(cause) =>
+            onFailed.invoke(cause)
+        })(ExecutionContexts.parasitic)
+
+        var requestOutPulled = false
+        requestOut.setHandler(new OutHandler {
+          override def onPull(): Unit =
+            requestOutPulled = true
+          override def onDownstreamFinish(): Unit = ()
+        })
+        responseIn.setHandler(new InHandler {
+          override def onPush(): Unit = ()
+          override def onUpstreamFinish(): Unit = ()
+          override def onUpstreamFailure(ex: Throwable): Unit = ()
+        })
+
+        override def onPush(): Unit = ()
+
+        override def onPull(): Unit = {
+          if (!isAvailable(requestIn) && !hasBeenPulled(requestIn)) // requestIn might already have been pulled when we failed and went back to Unconnected
+            pull(requestIn)
+        }
+
+        val onConnected = getAsyncCallback[Unit] { (_) =>
+          val newState = new Connected(requestOut, responseIn)
+          become(newState)
+          if (requestOutPulled) {
+            if (isAvailable(requestIn)) newState.dispatchRequest(grab(requestIn))
+            else if (!hasBeenPulled(requestIn)) pull(requestIn)
+          }
+        }
+        val onFailed = getAsyncCallback[Throwable] { cause =>
+          responseIn.cancel()
+          requestOut.fail(new RuntimeException("connection broken"))
+
+          if (connectsLeft.contains(0)) {
+            failStage(cause)
+          } else {
+            setHandler(requestIn, Unconnected)
+            log.info(s"failed, trying to connect again: ${cause.getMessage}${connectsLeft.map(n => s" ($n left)").getOrElse("")}")
```

Comment:
(another PR) 

Maybe we can add some more information here so user know what is the connection that's failing (the user may have multiple clients, or is using service discovery and doesn't have access to the resolved host:port value or...)